packages/core/src/external/mpi/match-patients.ts

```diff
@@ -0,0 +1,147 @@
+type SimilarityFunction = (
+  patient1: PatientDataMPI,
+  patient2: PatientDataMPI,
+  threshold: number
+) => boolean;
+
+type MatchingRule = (patient1: PatientDataMPI, patient2: PatientDataMPI) => boolean;
+
+/**
+ * `matchPatients` filters patients based on a similarity function, rules, and a threshold.
+ * @param similarityFunction - Determines if a patient is a match.
+ * @param matchingRules - Rules that determines if a patient is a match.
+ * @param patients - Array of patients.
+ * @param demo - Patient data to find matches for.
+ * @param threshold - Minimum similarity score for a match.
+ * @param greedy - If true, returns the first match. If false, returns all matches.
+ * @returns matched patients.
+ */
+export const matchPatients = (
+  similarityFunction: SimilarityFunction,
+  matchingRules: MatchingRule[],
+  patients: PatientDataMPI[],
+  demo: PatientDataMPI,
+  threshold: number,
+  greedy = true
+): PatientDataMPI[] => {
+  const matchFunction = (patient: PatientDataMPI) => {
+    const PatientDataMPI = normalizePatientDataMPI(patient);
+    if (!PatientDataMPI) {
+      return false;
+    }
+    for (const rule of matchingRules) {
+      if (rule(demo, PatientDataMPI)) {
+        return true;
+      }
+    }
+    return similarityFunction(PatientDataMPI, demo, threshold);
+  };
+  if (greedy) {
+    const foundPatient = patients.find(matchFunction);
+    return foundPatient ? [foundPatient] : [];
+  } else {
+    return patients.filter(matchFunction);
+  }
+};
+
+/**
+ * This function checks if the patient has any personal identifiers that match the demo. The idea of having rules
+ * is that rules will override statistical similarity methods. This rule should act as an override.
+ * @param demo
+ * @param patient
+ * @returns true if the patient has any personal identifiers that match the demo.
+ */
+export const matchingPersonalIdentifiersRule = (
+  demo: PatientDataMPI,
+  patient: PatientDataMPI
+): boolean => {
+  return Boolean(
+    intersectionWith(patient.personalIdentifiers || [], demo.personalIdentifiers || [], isEqual)
+      .length > 0
+  );
+};
+
+/**
+ * This function checks if the patient has any contact details (phone or email) that match the demo. The idea of this rule
+ * us that contact info is a very strong unique identifier if they do match, but doesn't necessarily mean that the patients
+ * don't match if they don't match.
+ * @param demo
+ * @param patient
+ * @returns true if the patient has any contact details that match the demo.
+ */
+export const matchingContactDetailsRule = (
+  demo: PatientDataMPI,
+  patient: PatientDataMPI
+): boolean => {
+  // Boolean so if undefined || undefined, result will be false and not just falsy
+  return Boolean(
+    intersectionWith(patient.contact || [], demo.contact || [], (a, b) => a?.phone === b?.phone)
+      .length > 0 ||
+      intersectionWith(patient.contact || [], demo.contact || [], (a, b) => a?.email === b?.email)
+        .length > 0
+  );
+};
```

Comment:
I suggest we create small functions and constants to simplify the code:

```ts
function isSameContactByPhone(a?: Contact, b?: Contact): boolean {
   return a?.phone === b?.phone;
}
function isSameContactByEmail(a?: Contact, b?: Contact): boolean {
   return a?.email === b?.email;
}
function...
  const contact1 = patient1.contact || [];
  const contact1 = patient2.contact || [],
  const isMatchPhone = intersectionWith(contact1, contact2, isSameContactByPhone).length > 0;
  const isMatchEmail = intersectionWith(contact1, contact2, isSameContactByEmail).length > 0;
  return isMatchPhone || isMatchEmail
```

And why the use of `Boolean()`?

I would also apply this to the other match function(s)